# pandas 실습 - 데이터 특성을 알아보자

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. pandas와 numpy 라이브러리를 Import 합니다.

In [10]:
url = "https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv"
import pandas as pd

pd.read_csv(url, delimiter='\t').to_csv('chipotle.tsv', sep="\t", index=False)

In [21]:
import pandas as pd
import numpy as np

### Step 2. 데이터 셋을 불러옵니다.

In [13]:
chipo = pd.read_csv("chipotle.tsv", sep = '\t')
chipo["item_price2"] = chipo['item_price'].apply(lambda x : float(x[1:]) )
chipo

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  \
0                                                   NaN     $2.39    
1                                          [Clementine]     $3.39    
2                                               [Apple]     $3.39    
3                                                   NaN     $2.39    
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98    
...                                                 ...        ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75    
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75    
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25    
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75    
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75    

      item_price2  
0            2.39  
1            3.39  
2            3.39  
3            2.39  
4           16.98  
...           ...  
4617        11.75  
4618        11.75  
4619        11.25  
4620         8.75  
4621         8.75  

[4622 rows x 6 columns]

### Step 3. head 함수를 통해 처음 10개 행을 살펴봅시다.

In [15]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  item_price2  
0                                                NaN     $2.39          2.39  
1                                       [Clementine]     $3.39          3.39  
2                                            [Apple]     $3.39          3.39  
3                                                NaN     $2.39          2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98         16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98         10.98  
6                                                NaN     $1.69          1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75         11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25          9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25          9.25

In [102]:
import numpy as np

mean_and_round = lambda x: round(np.mean(x), 2)

grouped_df = chipo.groupby('item_name').agg({
    "quantity": [ sum, max ],
    'item_price2': [ ('M&R', mean_and_round), ('sum', sum) ],
}).sort_values(('quantity', 'sum'), ascending=False)


grouped_df

quantity     item_price2         
                                           sum max         M&R      sum
item_name                                                              
Chicken Bowl                               761   3       10.11  7342.73
Chicken Burrito                            591   4       10.08  5575.82
Chips and Guacamole                        506   4        4.60  2201.04
Steak Burrito                              386   3       10.47  3851.43
Canned Soft Drink                          351   4        1.46   438.75
Chips                                      230   3        2.34   494.34
Steak Bowl                                 221   3       10.71  2260.19
Bottled Water                              211  10        1.87   302.56
Chips and Fresh Tomato Salsa               130  15        3.29   361.36
Canned Soda                                126   4        1.32   137.34
Chicken Salad Bowl                         123   2       11.17  1228.75
Chicken Soft Tacos                         120   2        9.64  1108.09
Side of Chips                              110   8        1.84   185.90
Veggie Burrito                              97   3        9.84   934.77
Barbacoa Burrito                            91   1        9.83   894.75
Veggie Bowl                                 87   2       10.21   867.99
Carnitas Bowl                               71   3       10.83   736.71
Barbacoa Bowl                               66   1       10.19   672.36
Carnitas Burrito                            60   2       10.13   597.83
Steak Soft Tacos                            56   2        9.75   536.05
6 Pack Soft Drink                           55   2        6.61   356.95
Chips and Tomatillo Red Chili Salsa         50   2        3.07   147.50
Chicken Crispy Tacos                        50   2       10.05   472.13
Chips and Tomatillo Green Chili Salsa       45   2        3.09   132.75
Carnitas Soft Tacos                         40   1        9.40   375.94
Steak Crispy Tacos                          36   2       10.21   357.34
Chips and Tomatillo-Green Chili Salsa       33   2        2.54    78.87
Steak Salad Bowl                            31   2       11.85   343.59
Nantucket Nectar                            29   2        3.64    98.31
Chips and Tomatillo-Red Chili Salsa         25   3        2.99    59.75
Barbacoa Soft Tacos                         25   1       10.02   250.46
Chips and Roasted Chili Corn Salsa          23   2        3.08    67.85
Izze                                        20   1        3.39    67.80
Veggie Salad Bowl                           18   1       10.14   182.50
Chips and Roasted Chili-Corn Salsa          18   1        2.39    43.02
Barbacoa Crispy Tacos                       12   2       10.93   120.21
Barbacoa Salad Bowl                         10   1       10.64   106.40
Chicken Salad                                9   1        9.01    81.09
Carnitas Crispy Tacos                        8   2       11.14    77.96
Veggie Soft Tacos                            8   2       10.57    73.96
Burrito                                      6   1        7.40    44.40
Veggie Salad                                 6   1        8.49    50.94
Carnitas Salad Bowl                          6   1       11.06    66.34
Bowl                                         4   3       14.80    29.60
Steak Salad                                  4   1        8.91    35.66
Salad                                        2   1        7.40    14.80
Crispy Tacos                                 2   1        7.40    14.80
Chips and Mild Fresh Tomato Salsa            1   1        3.00     3.00
Carnitas Salad                               1   1        8.99     8.99
Veggie Crispy Tacos                          1   1        8.49     8.49

quantity     item_price2           \
                                           sum max        mean      sum   
item_name                                                                 
Chicken Bowl                               761   3   10.113953  7342.73   
Chicken Burrito                            591   4   10.082857  5575.82   
Steak Burrito                              386   3   10.465842  3851.43   
Steak Bowl                                 221   3   10.711801  2260.19   
Chips and Guacamole                        506   4    4.595073  2201.04   
Chicken Salad Bowl                         123   2   11.170455  1228.75   
Chicken Soft Tacos                         120   2    9.635565  1108.09   
Veggie Burrito                              97   3    9.839684   934.77   
Barbacoa Burrito                            91   1    9.832418   894.75   
Veggie Bowl                                 87   2   10.211647   867.99   
Carnitas Bowl                               71   3   10.833971   736.71   
Barbacoa Bowl                               66   1   10.187273   672.36   
Carnitas Burrito                            60   2   10.132712   597.83   
Steak Soft Tacos                            56   2    9.746364   536.05   
Chips                                      230   3    2.342844   494.34   
Chicken Crispy Tacos                        50   2   10.045319   472.13   
Canned Soft Drink                          351   4    1.457641   438.75   
Carnitas Soft Tacos                         40   1    9.398500   375.94   
Chips and Fresh Tomato Salsa               130  15    3.285091   361.36   
Steak Crispy Tacos                          36   2   10.209714   357.34   
6 Pack Soft Drink                           55   2    6.610185   356.95   
Steak Salad Bowl                            31   2   11.847931   343.59   
Bottled Water                              211  10    1.867654   302.56   
Barbacoa Soft Tacos                         25   1   10.018400   250.46   
Side of Chips                              110   8    1.840594   185.90   
Veggie Salad Bowl                           18   1   10.138889   182.50   
Chips and Tomatillo Red Chili Salsa         50   2    3.072917   147.50   
Canned Soda                                126   4    1.320577   137.34   
Chips and Tomatillo Green Chili Salsa       45   2    3.087209   132.75   
Barbacoa Crispy Tacos                       12   2   10.928182   120.21   
Barbacoa Salad Bowl                         10   1   10.640000   106.40   
Nantucket Nectar                            29   2    3.641111    98.31   
Chicken Salad                                9   1    9.010000    81.09   
Chips and Tomatillo-Green Chili Salsa       33   2    2.544194    78.87   
Carnitas Crispy Tacos                        8   2   11.137143    77.96   
Veggie Soft Tacos                            8   2   10.565714    73.96   
Chips and Roasted Chili Corn Salsa          23   2    3.084091    67.85   
Izze                                        20   1    3.390000    67.80   
Carnitas Salad Bowl                          6   1   11.056667    66.34   
Chips and Tomatillo-Red Chili Salsa         25   3    2.987500    59.75   
Veggie Salad                                 6   1    8.490000    50.94   
Burrito                                      6   1    7.400000    44.40   
Chips and Roasted Chili-Corn Salsa          18   1    2.390000    43.02   
Steak Salad                                  4   1    8.915000    35.66   
Bowl                                         4   3   14.800000    29.60   
Salad                                        2   1    7.400000    14.80   
Crispy Tacos                                 2   1    7.400000    14.80   
Carnitas Salad                               1   1    8.990000     8.99   
Veggie Crispy Tacos                          1   1    8.490000     8.49   
Chips and Mild Fresh Tomato Salsa            1   1    3.000000     3.00   

                                      item_price3  
          

In [72]:
pivot_df = chipo.pivot_table(index="order_id", values="quantity", columns="item_name", aggfunc="sum").fillna(0)
pivot_df.to_csv("result.csv",encoding="cp949")

In [30]:
chipo["item_price2"].sum()

34500.16

In [39]:
adf = adf.sort_values("quantity", ascending=False)
adf["profit"] = adf.apply(lambda x: x["quantity"] * x["item_price_2"], axis=1)

In [42]:
adf.sort_values("profit", ascending=False)

quantity  item_price_2       profit
item_name                                               
Chicken Bowl              761     10.113953  7696.718361
Chicken Burrito           591     10.082857  5958.968571
Steak Burrito             386     10.465842  4039.815163
Steak Bowl                221     10.711801  2367.308009
Chips and Guacamole       506      4.595073  2325.106973

# 500개 미만 판매량을 가진 메뉴 중에서 총 매출이 2000$ 이상인 행만 추려봅시다.

In [45]:
adf[ (adf["quantity"] < 500) & (adf["profit"] >= 2000)]

quantity  item_price_2       profit
item_name                                         
Steak Burrito       386     10.465842  4039.815163
Steak Bowl          221     10.711801  2367.308009

#  또는 수량이 50 이하인 행

In [55]:
adf[ ((adf["quantity"] < 500) & (adf["profit"] >= 2000)) | (adf["quantity"] < 50) ]

quantity  item_price_2       profit
item_name                                                                 
Steak Burrito                               386     10.465842  4039.815163
Steak Bowl                                  221     10.711801  2367.308009
Chips and Tomatillo Green Chili Salsa        45      3.087209   138.924419
Carnitas Soft Tacos                          40      9.398500   375.940000
Steak Crispy Tacos                           36     10.209714   367.549714
Chips and Tomatillo-Green Chili Salsa        33      2.544194    83.958387
Steak Salad Bowl                             31     11.847931   367.285862
Nantucket Nectar                             29      3.641111   105.592222
Chips and Tomatillo-Red Chili Salsa          25      2.987500    74.687500
Barbacoa Soft Tacos                          25     10.018400   250.460000
Chips and Roasted Chili Corn Salsa           23      3.084091    70.934091
Izze                                         20      3.390000    67.800000
Veggie Salad Bowl                            18     10.138889   182.500000
Chips and Roasted Chili-Corn Salsa           18      2.390000    43.020000
Barbacoa Crispy Tacos                        12     10.928182   131.138182
Barbacoa Salad Bowl                          10     10.640000   106.400000
Chicken Salad                                 9      9.010000    81.090000
Veggie Soft Tacos                             8     10.565714    84.525714
Carnitas Crispy Tacos                         8     11.137143    89.097143
Burrito                                       6      7.400000    44.400000
Veggie Salad                                  6      8.490000    50.940000
Carnitas Salad Bowl                           6     11.056667    66.340000
Bowl                                          4     14.800000    59.200000
Steak Salad                                   4      8.915000    35.660000
Salad                                         2      7.400000    14.800000
Crispy Tacos                                  2      7.400000    14.800000
Chips and Mild Fresh Tomato Salsa             1      3.000000     3.000000
Carnitas Salad                                1      8.990000     8.990000
Veggie Crispy Tacos                           1      8.490000     8.490000

In [70]:
chipo["item_name"].sort_values().unique()

array(['6 Pack Soft Drink', 'Barbacoa Bowl', 'Barbacoa Burrito',
       'Barbacoa Crispy Tacos', 'Barbacoa Salad Bowl',
       'Barbacoa Soft Tacos', 'Bottled Water', 'Bowl', 'Burrito',
       'Canned Soda', 'Canned Soft Drink', 'Carnitas Bowl',
       'Carnitas Burrito', 'Carnitas Crispy Tacos', 'Carnitas Salad',
       'Carnitas Salad Bowl', 'Carnitas Soft Tacos', 'Chicken Bowl',
       'Chicken Burrito', 'Chicken Crispy Tacos', 'Chicken Salad',
       'Chicken Salad Bowl', 'Chicken Soft Tacos', 'Chips',
       'Chips and Fresh Tomato Salsa', 'Chips and Guacamole',
       'Chips and Mild Fresh Tomato Salsa',
       'Chips and Roasted Chili Corn Salsa',
       'Chips and Roasted Chili-Corn Salsa',
       'Chips and Tomatillo Green Chili Salsa',
       'Chips and Tomatillo Red Chili Salsa',
       'Chips and Tomatillo-Green Chili Salsa',
       'Chips and Tomatillo-Red Chili Salsa', 'Crispy Tacos', 'Izze',
       'Nantucket Nectar', 'Salad', 'Side of Chips', 'Steak Bowl',
       'Steak

# chicken이라는 단어가 들어간 메뉴의 가지수는 얼마나 될까?

In [37]:
chipo[ chipo["item_name"].str.contains("(?i)chicken") ]

order_id  quantity             item_name  \
4            2         2          Chicken Bowl   
5            3         1          Chicken Bowl   
11           6         1  Chicken Crispy Tacos   
12           6         1    Chicken Soft Tacos   
13           7         1          Chicken Bowl   
...        ...       ...                   ...   
4604      1828         1          Chicken Bowl   
4615      1832         1    Chicken Soft Tacos   
4619      1834         1    Chicken Salad Bowl   
4620      1834         1    Chicken Salad Bowl   
4621      1834         1    Chicken Salad Bowl   

                                     choice_description item_price  \
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98    
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98    
11    [Roasted Chili Corn Salsa, [Fajita Vegetables,...     $8.75    
12    [Roasted Chili Corn Salsa, [Rice, Black Beans,...     $8.75    
13    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...    $11.25    
...                                                 ...        ...   
4604  [Fresh Tomato Salsa, [Rice, Black Beans, Chees...     $8.75    
4615   [Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]     $8.75    
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25    
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75    
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75    

      item_price2  
4           16.98  
5           10.98  
11           8.75  
12           8.75  
13          11.25  
...           ...  
4604         8.75  
4615         8.75  
4619        11.25  
4620         8.75  
4621         8.75  

[1560 rows x 6 columns]

In [ ]:
chipo[ chipo["item_name"].str.contains("(?i)chicken") ]

# Cicken와 Salad가 모두 들어간 메뉴의 가지수는 얼마나 될까?

In [45]:
filtered = chipo[ (chipo["item_name"].str.contains("(?i)chicken")) & (chipo["item_name"].str.contains("(?i)salad")) ]
filtered["item_name"].unique()

array(['Chicken Salad Bowl', 'Chicken Salad'], dtype=object)

In [92]:
ddf = chipo[ chipo["item_name"].str.contains("Chicken") | chipo["item_name"].str.contains("Salsa") ]
ddf["item_name"].unique().shape[0]

14

# 아래와 같이 각 주문에 대한 주문자 이름의 DF가 있을때, 주문데이터와 left 조인 해봅시다.

In [78]:
orderer_name = [
    [1, "louis"],
    [2, "nancy"],
    [3, "bob"],
    [5500, "alice"]
]

columns = [ "order_id", "orderer_name"]

name_df = pd.DataFrame(orderer_name, columns=columns)
name_df = name_df.set_index("order_id")
name_df

orderer_name
order_id             
1               louis
2               nancy
3                 bob
5500            alice

In [79]:
pd.merge(name_df,chipo, how="inner", left_index=True, right_on="order_id").head(10)

orderer_name  order_id  quantity                              item_name  \
0        louis         1         1           Chips and Fresh Tomato Salsa   
1        louis         1         1                                   Izze   
2        louis         1         1                       Nantucket Nectar   
3        louis         1         1  Chips and Tomatillo-Green Chili Salsa   
4        nancy         2         2                           Chicken Bowl   
5          bob         3         1                           Chicken Bowl   
6          bob         3         1                          Side of Chips   

                                  choice_description item_price  item_price2  
0                                                NaN     $2.39          2.39  
1                                       [Clementine]     $3.39          3.39  
2                                            [Apple]     $3.39          3.39  
3                                                NaN     $2.39          2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98         16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98         10.98  
6                                                NaN     $1.69          1.69

order_id orderer_name  quantity                              item_name  \
0         1        louis         1           Chips and Fresh Tomato Salsa   
1         1        louis         1                                   Izze   
2         1        louis         1                       Nantucket Nectar   
3         1        louis         1  Chips and Tomatillo-Green Chili Salsa   
4         2        nancy         2                           Chicken Bowl   
5         3          bob         1                           Chicken Bowl   
6         3          bob         1                          Side of Chips   

                                  choice_description item_price  
0                                               정보없음     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                               정보없음     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                               정보없음     $1.69

In [9]:
chipo["item_price_2"] = chipo["item_price"].apply(lambda x: float(x[1:]) )
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  item_price_2  
0                                                NaN     $2.39           2.39  
1                                       [Clementine]     $3.39           3.39  
2                                            [Apple]     $3.39           3.39  
3                                                NaN     $2.39           2.39  
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98          16.98  
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98          10.98  
6                                                NaN     $1.69           1.69  
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75          11.75  
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25           9.25  
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25           9.25

In [33]:
adf = chipo.groupby(by="item_name").agg({
    "quantity":"sum",
    "item_price_2": "mean"
})
adf = adf.sort_values('quantity',ascending=False)
adf

KeyError: "Column(s) ['item_price_2'] do not exist"

In [12]:
adf = chipo.groupby(by="item_name").sum().drop("order_id",axis=1)
adf.sort_values("quantity", ascending=False)

quantity  item_price_2
item_name                                                    
Chicken Bowl                                761       7342.73
Chicken Burrito                             591       5575.82
Chips and Guacamole                         506       2201.04
Steak Burrito                               386       3851.43
Canned Soft Drink                           351        438.75
Chips                                       230        494.34
Steak Bowl                                  221       2260.19
Bottled Water                               211        302.56
Chips and Fresh Tomato Salsa                130        361.36
Canned Soda                                 126        137.34
Chicken Salad Bowl                          123       1228.75
Chicken Soft Tacos                          120       1108.09
Side of Chips                               110        185.90
Veggie Burrito                               97        934.77
Barbacoa Burrito                             91        894.75
Veggie Bowl                                  87        867.99
Carnitas Bowl                                71        736.71
Barbacoa Bowl                                66        672.36
Carnitas Burrito                             60        597.83
Steak Soft Tacos                             56        536.05
6 Pack Soft Drink                            55        356.95
Chips and Tomatillo Red Chili Salsa          50        147.50
Chicken Crispy Tacos                         50        472.13
Chips and Tomatillo Green Chili Salsa        45        132.75
Carnitas Soft Tacos                          40        375.94
Steak Crispy Tacos                           36        357.34
Chips and Tomatillo-Green Chili Salsa        33         78.87
Steak Salad Bowl                             31        343.59
Nantucket Nectar                             29         98.31
Chips and Tomatillo-Red Chili Salsa          25         59.75
Barbacoa Soft Tacos                          25        250.46
Chips and Roasted Chili Corn Salsa           23         67.85
Izze                                         20         67.80
Veggie Salad Bowl                            18        182.50
Chips and Roasted Chili-Corn Salsa           18         43.02
Barbacoa Crispy Tacos                        12        120.21
Barbacoa Salad Bowl                          10        106.40
Chicken Salad                                 9         81.09
Carnitas Crispy Tacos                         8         77.96
Veggie Soft Tacos                             8         73.96
Burrito                                       6         44.40
Veggie Salad                                  6         50.94
Carnitas Salad Bowl                           6         66.34
Bowl                                          4         29.60
Steak Salad                                   4         35.66
Salad                                         2         14.80
Crispy Tacos                                  2         14.80
Chips and Mild Fresh Tomato Salsa             1          3.00
Carnitas Salad                                1          8.99
Veggie Crispy Tacos                           1          8.49

# choice_description 컬럼의 결측값을 채워봅시다
---


In [30]:
chipo

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description item_price  
0                                                   NaN     $2.39   
1                                          [Clementine]     $3.39   
2                                               [Apple]     $3.39   
3                                                   NaN     $2.39   
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
...                                                 ...        ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...    $11.75   
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...    $11.75   
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...    $11.25   
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...     $8.75   
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...     $8.75   

[4622 rows x 5 columns]

### Step 4. 데이터셋의 크기를 살펴봅시다.

In [10]:
# Solution 1

chipo.shape

(4622, 6)

In [10]:
chipo.index.size # index의 size로 살펴보는 방법

4622

In [5]:
chipo.info() # info 함수를 이용하는 방법

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
order_id              4622 non-null int64
quantity              4622 non-null int64
item_name             4622 non-null object
choice_description    3376 non-null object
item_price            4622 non-null object
dtypes: int64(2), object(3)
memory usage: 180.6+ KB


### Step 5. 데이터의 컬럼은 몇개인지 살펴봅시다.

In [6]:
chipo.shape[1]

5

### Step 6. 컬럼명들을 살펴봅시다.

In [7]:
chipo.columns

Index([u'order_id', u'quantity', u'item_name', u'choice_description',
       u'item_price'],
      dtype='object')

### Step 7. 데이터 셋은 어떻게 인덱싱 되어 있을까요?

In [8]:
chipo.index

RangeIndex(start=0, stop=4622, step=1)

0부터 4622의 연속된 정수(순번)로 인덱싱 되어 있습니다.

### Step 8. 가장 인기 많은 메뉴는 무엇일까요?

In [13]:
c = chipo.groupby('item_name')
c = c.sum()
c = c.sort_values(['quantity'], ascending=False)
c.head(1)

order_id  quantity
item_name                       
Chicken Bowl    713926       761

### Step 9. 가장 많이 선택된 사이드 메뉴(choice_description)는 무엇일까요?

In [15]:
c = chipo.groupby('choice_description').sum()
c = c.sort_values(['quantity'], ascending=False)
c.head(1)
# Diet Coke 159

order_id  quantity
choice_description                    
[Diet Coke]           123455       159

### Step 10. 총 주문 수를 구해봅시다.

In [16]:
total_items_orders = chipo.quantity.sum()
total_items_orders

4972

### Step 11. 정수값의 price 컬럼을 float으로 바꾸어 봅시다.

#### Step 11.a. Check the item price type

In [17]:
chipo.item_price.dtype

dtype('O')

#### Step 11.b. 람다를 활용해 문자열을 float으로 바꾸어 봅시다.

In [20]:
dollarizer = lambda x: float(x[1:-1])
chipo.item_price = chipo.item_price.apply(dollarizer)

order_id  quantity                              item_name  \
0            1         1           Chips and Fresh Tomato Salsa   
1            1         1                                   Izze   
2            1         1                       Nantucket Nectar   
3            1         1  Chips and Tomatillo-Green Chili Salsa   
4            2         2                           Chicken Bowl   
...        ...       ...                                    ...   
4617      1833         1                          Steak Burrito   
4618      1833         1                          Steak Burrito   
4619      1834         1                     Chicken Salad Bowl   
4620      1834         1                     Chicken Salad Bowl   
4621      1834         1                     Chicken Salad Bowl   

                                     choice_description  item_price  
0                                                   NaN        2.39  
1                                          [Clementine]        3.39  
2                                               [Apple]        3.39  
3                                                   NaN        2.39  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
...                                                 ...         ...  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...       11.25  
4620  [Fresh Tomato Salsa, [Fajita Vegetables, Lettu...        8.75  
4621  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...        8.75  

[4622 rows x 5 columns]

#### Step 11.c. 바뀐 컬럼의 타입을 확인해 보기.

In [23]:
chipo.item_price.dtype

dtype('float64')

### Step 12. 총 수입 금액을 구해봅시다.

In [24]:
revenue = (chipo['quantity']* chipo['item_price']).sum()

print('Revenue was: $' + str(np.round(revenue,2)))

Revenue was: $39237.02


### Step 13. 각 주문의 평균 금액을 구해 봅시다.

In [30]:
# Solution 1

chipo['revenue'] = chipo['quantity'] * chipo['item_price']
order_grouped = chipo.groupby(by=['order_id']).sum()
order_grouped.mean()['revenue']

21.39423118865867

In [28]:
# Solution 2

chipo.groupby(by=['order_id']).sum().mean()['revenue']

21.39423118865867

### Step 14. 메뉴의 가지수를 구해봅시다.

In [29]:
chipo.item_name.value_counts().count()

50